lighthouse-core/gather/gatherers/seo/tap-targets.js

```diff
@@ -0,0 +1,419 @@
+//   const {display, overflowX, overflowY} = style;
+
+//   if (display === 'inline' && overflowX === 'visible' && overflowY === 'visible') {
+//     const range = document.createRange();
+//     range.selectNodeContents(node);
+
+//     return range.getBoundingClientRect();
+//   }
+
+//   return node.getBoundingClientRect();
+// }
+
+function allBCRsEmpty(bcrs) {
+  return (
+    bcrs.length === 0 || bcrs.every(bcr => bcr.width === 0 && bcr.height === 0)
+  );
+}
+
+/**
+ * @param {Element} node
+ */
+function isVisible({node, bcrs = getBCRs(node, false)}) {
+  const {
+    overflowX,
+    overflowY,
+    display,
+    opacity,
+    visibility,
+  } = getComputedStyle(node);
+
+  if (
+    ((overflowX === 'hidden' && overflowY === 'hidden') ||
+      node.children.length === 0) &&
+    allBCRsEmpty(bcrs)
+  ) {
+    return false;
+  }
+
+  if (
+    display === 'none' ||
+    visibility === 'hidden' ||
+    visibility === 'collapse' ||
+    parseFloat(opacity) < 0.1
+  ) {
+    return false;
+  }
+
+  if (display === 'block' || display === 'inline-block') {
+    if (node.clientWidth === 0 && overflowX === 'hidden') {
+      return false;
+    }
+    if (node.clientHeight === 0 && overflowY === 'hidden') {
+      return false;
+    }
+  }
+
+  const parent = node.parentElement;
+
+  if (parent && parent.nodeName !== 'HTML' && !isVisible({node: parent})) {
+    return false;
+  }
+
+  return true;
+}
+// old isVisible code, code above is from gist
+// function isVisible(node) {
+//   const style = getComputedStyle(node);
+//   const {width, height} = getInteractiveBCR(node, style);
+//   const {overflowX, overflowY, display, opacity, visibility} = style;
+
+//   if (
+//     (width === 0 || height === 0) &&
+//     overflowX === 'hidden' &&
+//     overflowY === 'hidden') {
+//     return false;
+//   }
+
+//   if (
+//     display === 'none' ||
+//     visibility === 'hidden' ||
+//     visibility === 'collapse' ||
+//     // @ts-ignore opacity is reported as string|null while it can't be null
+//     parseFloat(opacity) < 0.1) {
+//     return false;
+//   }
+
+//   const parent = node.parentElement;
+
+//   if (
+//     parent &&
+//     parent.nodeName !== 'HTML' &&
+//     !isVisible(parent)
+//   ) {
+//     return false;
+//   }
+
+//   return true;
+// }
+
+/**
+ * @param {string} str
+ * @param {number} maxLength
+ * @returns {string}
+ */
+function truncate(str, maxLength) {
+  if (str.length <= maxLength) {
+    return str;
+  }
+  return str.slice(0, maxLength - 1) + '…';
+}
+
+function getBCRs(node, includeChildren = true) {
+  const {display, overflowX, overflowY} = getComputedStyle(node);
+
+  let bcrs;
+
+  // if (display === 'inline' && overflowX === 'visible' && overflowY === 'visible') {
+  //   const range = document.createRange();
+  //   range.selectNodeContents(node);
+
+  //   bcrs = range.getClientRects();
+  // } else {
+  bcrs = node.getClientRects();
```

Comment:
Yeah we could use getBCR – having a tap target split over multiple lines should be super rare. I don't think that would simplify the logic much though.

There are two bits of information we need for the overlap check:
1. where to put the finger on a tap area
2. where tap areas are

For (2) we should use client rects because it prevents false failures in cases like this:

![screen shot 2018-11-08 at 08 10 16](https://user-images.githubusercontent.com/1303660/48199827-3a9ea180-e355-11e8-94b5-5a6405eb773a.png)

We can't just check the tap target element itself but also need to check its children, because they can be outside the parent BCR. For example, images inside of an `<a>` tag:

![screen shot 2018-11-08 at 08 44 25](https://user-images.githubusercontent.com/1303660/48200130-1a231700-e356-11e8-89fd-ebf2ed2bf4b4.png)

So even with getBCR we end up with multiple client rects, but we don't want to put a finger on each one. So we merge them together. (In the case above simplifyBCRs merges `<a>` and `<img>` BCRs together.)

Another example that gets merged:

![screen shot 2018-11-08 at 10 11 52](https://user-images.githubusercontent.com/1303660/48199971-9e28cf00-e355-11e8-8cbf-fdbef9d5cbe3.png)

We also can't just get the bounding rect of all client rects, because some of the tap target children might, for example, be `position: absolute`.

Reply:
Thanks for the details and example this helps a lot! Few more questions :)

I'm not sure I totally understood what the `(2)` case was with the "Bizarre penalty decision" example. Are those the green boxes if we use client rects, and gray box is getBCR? Is the issue that the european football finger would bleed into the getBCR box but not getClientRects one?

> but we don't want to put a finger on each one

Just thinking aloud here, what if we convert this constraint a little bit. Right now our logic is basically "if there is a place to tap the target unambiguously, then it passes" (i.e. if you can put your finger in the middle). Given this policy, we should be treating every child outside the BCR of the parent as a new potential tap target. It also looks like the current situation with  extending the BCR into non-rectangular shapes could also lead to putting our finger tap completely outside the bounds of the merged shape. 

What happens if we treated them separately and de-duped them at the end? I'm trying to see if there's a consistent way to avoid the complicated BCR merging. I think it makes the audit hard to reason about and explaining/understanding failures will be difficult to users as well. I think you already know very well--and as I'm starting to suspect--it's impossible to have this not be complicated and confusing, but I at least want to give it my best shot :)

The biggest sticking point to me seems to be knowing that tap targets will lead to the same result. We are using shared parent elements as a proxy for this, but we don't really know for sure. Is that a fair assessment?

Reply:
> I'm trying to see if there's a consistent way to avoid the complicated BCR merging. I think it makes the audit hard to reason about and explaining/understanding failures will be difficult to users as well.

Agreed. Would be great to find an alternative solution or some trade-off we can make 🙂.

> Are those the green boxes if we use client rects, and gray box is getBCR? Is the issue that the european football finger would bleed into the getBCR box but not getClientRects one?

Yes to both questions.

> The biggest sticking point to me seems to be knowing that tap targets will lead to the same result. We are using shared parent elements as a proxy for this, but we don't really know for sure (Example "Senator reacts" one, do both of those go to the same place? Does the play button take you somewhere different? Can we always tell?). Is that a fair assessment?

No, the issue here is different. There's only one `<a>` tag, which contains some text and the play button. We know they have the same action, the actual issue is with children outside the parent BCR.

We don't report a failure if one tap target is completely inside another one, because that's usually an intentional design decision.

> Given this policy, we should be treating every child outside the BCR of the parent as a new potential tap target.

We used to do that, but changed it because it was failing cases that shouldn't fail:

![screen shot 2018-11-08 at 16 15 50](https://user-images.githubusercontent.com/1303660/48211547-9bd46e00-e371-11e8-8e40-1c603e62efdb.png)

It might be a good idea to pass every element that has at least one unambiguous tap target finger area. That would fail rects that are tall enough collectively but not on their own, though.

Reply:
Update after talking with @rviscomi:

We're concerned about not reporting what should be failures when using the "at least one non-overlapping tap finger" approach. For example, if there's another target right next to the "senators react" link but it doesn't reach all the way to the play button then the tap target should fail. We don't want to allow the small extra rect/finger to count as a fix to the problem.

We can fix the "finger tap completely outside the bounds of the merged shape" problem by checking that the center of the merged area is within one of the input rects, and if not cancel the merge.

Reply:
Alright, gotcha that makes sense! But then I think we should reconsider the center of rect approach then if we're wanting to flag anything that's too close.

WDYT about...

A tap target is "too small" if the non-ambiguous tappable area is <X% of the total area.

Reply:
I like that idea! Is this what you mean?

![screen shot 2018-11-09 at 10 34 16](https://user-images.githubusercontent.com/1303660/48257670-0dacc600-e40b-11e8-9b17-05d83f381f1e.png)

(Dividing by own area might be better, and would also strongly penalize small tap targets.)

So this would be the steps:

1. For each client rect add 24px finger area on each side
2. Get total area of the merged polygons
3. Get total area of the merged polygons overlapping other targets

I'll implement it in the gist and see how well it works in practice.

Reply:
Two thoughts on the approach that's not finger-based:

- would have a high ambiguity score for a 47x47px target (it currently doesn't take into account that people will try to tap in the middle of a tap target)
- the `ambiguousAreaRects` or `clientRects` could overlap with each other... so ideally we'd merge them each into one polygon, get a union polygon where the two overlap, and then get the area of that polygon. Feels a bit tricky.
- I think it only avoids the multi-rect issue by considering more are for the score? Multiple small target rects would still cause a large finger area.

Fundamentally it's a good idea to consider the entire tappable area (rather than just the center), but in practice I don't think it really makes much difference. The user will try to tap in the center of a button (because otherwise they might accidentally tap on another button). And usually the surrounding tap targets at the center of a tap target are representative of the whole target.

Here's a screenshot from playing around with it, the score is how much of the target is overlapped by other target's extended area):

![screenshot 2018-11-13 at 16 23 10](https://user-images.githubusercontent.com/1303660/48427430-9b1c4d00-e760-11e8-811f-3ce29392310e.png)

--- 

Been trying to figure out how all this maps to what we want to achieve.

Problem: Users try to tap on a small tap target but accidentally tap on another target. (@rviscomi I hope this sounds about right 🙂)

A target passes if:
- it's at least 48x48px *or*
- it's too small but missing it doesn't accidentally trigger an unintended action

Complications:
- shape of tappable area isn't a single rectangle (though usually it almost is)
- deciding likelihood of accidentally triggering another action (we're ok with some finger area overlap)

Using the finger approach, these are our options for handling multiple rects:
- merge (and check all after merge)
- check largest rect only
- pass if any passes

Given that we'll never report a tap target that has a rect > 48x48px, and that multiple rects are uncommon, I was quite hopeful that we could avoid the merging. 

But actually there is a meaningful score impact. On the left is with the merge, on the right we pass if any client rect passes.

![screenshot 2018-11-13 at 17 12 46](https://user-images.githubusercontent.com/1303660/48430495-62cc3d00-e767-11e8-8833-e3563f3c3ca9.png)

"check largest rect" also would cause problems, for example with multiple lines of text (like the "Winds whip" headline above) where the longest also happens to be closest to other rects.

If we use "any rect passes" alongside merging we could use some slightly simpler merging logic (because we don't need to handle the `a` tag with image inside case). Haven't find any real world score differences between normal merging and simple merging+any rect passes.

Reply:
Thanks for investigating the alternatives so well! The audit wasn't my idea, so I'll definitely defer to the pre-established approach you think is best :)

I'm not 100% sure in the merging vs. non-merging example what new finger pads are causing the "Brexit: " lines to pass now. Naively they all seem to overlap with roughly the same amount of external client rect area as the original finger pads did.

Assuming I figure that one out though...

-------

Just to recap I think we've decided here the following:

1. Finger-based approach on the center of tap targets is what we'd like to go with.
2. We can't really avoid the merging of client rects.
3. Matt's done a ton more investigation than any developer should need to for a single PR and it's time for me to just review some code? @rviscomi are you happy with these decisions too :)

Reply:
> Assuming I figure that one out though...

That's just a bug in my code. I thought the reason it didn't fail is because of the extra empty space on the left of the tap target, but that doesn't apply because we detect the quarter finger with the largest overlap.

Reply:
@patrickhulce Should be good to take another look at the code now.